### PR TITLE
COMPONENT_DUPE_SELECTIVE

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
Adds new component dupe mode that allows selective duplicates, the existing duplicates each decide if they will absorb the new component or allow it to exist separately.

## Changelog
:cl:
code: adds selective duplicate component mode
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
what justification is there for adding this

**Comment by spookydonut** [comment]
> 
> 
> what justification is there for adding this

@ninjanomnom has expressed a desire for this previously

**Comment by Moja10** [comment]
Does this allow you duplicate mobs without fucking everything if you put a soul into the cloned body or am I clinically retarded?

File: code/datums/components/_component.dm [hunk 1]
```diff
@@ -390,6 +401,18 @@
 						old_comp.InheritComponent(null, TRUE, arguments)
 					else
 						old_comp.InheritComponent(new_comp, TRUE)
+				if(COMPONENT_DUPE_SELECTIVE)
+					if(!new_comp)
+						var/list/arguments = args.Copy(2)
```
Comment (ninjanomnom) [inline code/datums/components/_component.dm]: You can do something like
```
arguments = args.Copy()
arguments[1] = new_comp
```
And you don't need to duplicate the loop

**Review by ninjanomnom — CHANGES_REQUESTED** [review]
`GetComponent` should runtime with components of this type like `COMPONENT_DUPE_ALLOWED`

File: code/datums/components/_component.dm [hunk 2]
```diff
@@ -390,6 +401,13 @@
...
 					else
 						old_comp.InheritComponent(new_comp, TRUE)
+				if(COMPONENT_DUPE_SELECTIVE)
+					var/list/arguments = args.Copy()
+					arguments[1] = new_comp
+					for(var/i in GetComponents(new_type))
+						var/datum/component/C = i
+						if(C.CheckDupeComponent(arguments))
+							break
 		else if(!new_comp)
```
Comment (ninjanomnom) [inline code/datums/components/_component.dm]: You need to actually make a new component

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/_component.dm [hunk 3]
```diff
@@ -390,6 +401,15 @@
 						old_comp.InheritComponent(null, TRUE, arguments)
 					else
 						old_comp.InheritComponent(new_comp, TRUE)
+				if(COMPONENT_DUPE_SELECTIVE)
+					if(!new_comp)
+						new_comp = new nt(arglist(args))
```
Comment (ninjanomnom) [inline code/datums/components/_component.dm]: You should only make the new component if it fails the checks on the existing components

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**